docs/enhancements/extract-os-image-from-release.md

```diff
@@ -0,0 +1,101 @@
+  - Extract OS image from the fetched `machine-os-images`.
+    E.g. 
+    $ oc image extract quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:0fc11beeab34f7187f6dc38ae60fb8e3e9d52d04a9a62d9f4aafe7192a7a83c1 --registry-config ~/.pullsecret --path /:/tmp/iso
+ - Populate the image store with the extracted OS image (i.e. cache the image).
+ * The create minimal ISO template flow should inject to the grub configs a reference to /boot-artifacts image-service API (instead of an external URL to rootfs).
+
+#### Pull-Secret Handling
+
+When using oc to extract an image, a pull-secret is required. Therefore, we need a means to provide it in each deployment:
+* SaaS: the pull-secret of the service account used for OCM communication can be fetched using AMS [access_token API](https://api.openshift.com/?urls.primaryName=Accounts%20management%20service#/default/post_api_accounts_mgmt_v1_access_token). The pull-secret can be then stored in a vault and used as a vault-secret for the image-service consumption. We probably should add some automation for fetching the pull-secret from AMS and update it in the vault (in case it gets invalidated).
+* On-Prem: add PULL_SECRET to podman deploy [configmap](https://github.com/openshift/assisted-service/blob/master/deploy/podman/configmap.yml). As we need an available pull-secret on image-service startup, a 'system' pull-secret is required for on-prem deployment.
+* Operator: add PULL_SECRET to [AgentServiceConfigSpec](https://github.com/openshift/assisted-service/blob/8f7bc60886cbab86104063d54d2a0e99fd538734/api/v1beta1/agentserviceconfig_types.go#L54)
+* CI: add PULL_SECRET env var for image-service (i.e. as done for [test-infra](https://github.com/openshift/release/blob/0e401d7e2fac1aa2beb84b4f9787f385897218cf/ci-operator/step-registry/assisted/baremetal/setup/assisted-baremetal-setup-commands.sh#L94)). Or, if applicable, add a vault-secret for it.
+
+#### Changes in the Operator
+
+As release images are supplied by specifying them in ClusterImageSet CRDs, we need a mechanism to notify the image-service on changes. This is required for the image-service to cache the OS images correlated with the releases.
+- This can be done by adding a watch for ClusterImageSets. When one is added/deleted, the image-service deployment should reflect it by getting an updated release images list (RELEASE_IMAGES env var). I.e. the operator should handle a change in release images list and update the relevant deployments accordingly.
+ - Should consider how to address a conflict in case of manual update of release_images env var by the user (instead of using ClusterImageSet).
+
+### Open Questions
+
+* For future support of dynamic OCP releases in SaaS, need to consider if it's required for the image-service to dynamically download OS images. I.e. we can either just pre-cache a set of OS images, or perhaps add some API in the image-service to extract on demand an OS image.
```

Comment:
I think what we would have to change here is replace all the operator stuff with some way for the image service to get an update from assisted-service when a new release image is added.

The idea is that the assisted-service would continue to monitor cluster image sets (or expose an API to add new release images in the SaaS/REST case) and track what releases are available in the database. Then the image-service would query that information from assisted. Maybe we could set up some kind of http pub/sub mechanism to subscribe to updates.

Reply:
I think all the logic in the image-service would be effectively the same.
Extracting the iso from the release image would work the same way if it's done at startup or at runtime.